File: src/com/soulmeadows/ValidParenthesis.java
package com.soulmeadows;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;

// https://leetcode.com/problems/valid-parentheses/description/
public class ValidParenthesis {
    public boolean isValid(String s) {
        var map = Map.of(')', '(', '}', '{', ']', '[');
        var stack = new ArrayDeque<Character>();
        for(var c: s.toCharArray()){
            if(map.containsKey(c)){
                var lastChar = stack.isEmpty()? '#': stack.pop();
                if(lastChar != map.get(c)){
                    return false;
                }
            }else{
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}

File: python/valid_parenthesis.py
# https://leetcode.com/problems/valid-parentheses/description/
class Solution:
    def isValid(self, s: str) -> bool:
        map: dict[str, str] ={")":"(", "}":"{", "]":"["}
        stack:list[str] = []
        for char in s:
            if char in map:
                lastChar = stack.pop() if stack else "#"
                if lastChar != map[char]:
                    return False
            else:
                stack.append(char)
        return not stack
